### PR TITLE
PR: Migrate Main interpreter to new plugin API

### LINKED ISSUE #13103: Move Main Interpreter preferences to new Plugin API

### PR DESCRIPTION
**Pull request by goanpeca**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #13103

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @goanpeca

### CODE REVIEW

**Comment by goanpeca** [comment]
Hi @dalthviz, this one is ready for QA testing. Thanks!

**Comment by dalthviz** [comment]
Not sure why but seems like tests are failing on Windows @goanpeca (a message about Python aborting appears)

**Comment by goanpeca** [comment]
Hi @dalthviz 

>Not sure why but seems like tests are failing on Windows 

Yeah, I am not sure either, but please test the PR since the error seems to be random.

**Comment by dalthviz** [comment]
Seems like this is working with valid interpreters :+1: 

Also @goanpeca @ccordoba12 , while testing this I ran into an error where if you pass an invalid interpreter seems like it is being used by the LSP (?)

This happens not only in this PR but also in master and 4.x, although in 4.x no error report raises (but you can see that under the hood something is trying to execute the path given). I tested passing as interpreter a path to an installer (GIMP installer in this case):

```python

  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pyls_jsonrpc\endpoint.py", line 113, in consume
    self._handle_request(message['id'], message['method'], message.get('params'))
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pyls_jsonrpc\endpoint.py", line 182, in _handle_request
    handler_result = handler(params)
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pyls_jsonrpc\dispatchers.py", line 23, in handler
    return method(**(params or {}))
  File "c:\users\daniel\documents\spyder\spyder\external-deps\python-language-server\pyls\python_ls.py", line 334, in m_text_document__document_symbol
    return self.document_symbols(textDocument['uri'])
  File "c:\users\daniel\documents\spyder\spyder\external-deps\python-language-server\pyls\python_ls.py", line 250, in document_symbols
    return flatten(self._hook('pyls_document_symbols', doc_uri))
  File "c:\users\daniel\documents\spyder\spyder\external-deps\python-language-server\pyls\python_ls.py", line 156, in _hook
    return hook_handlers(config=self.config, workspace=workspace, document=doc, **kwargs)
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\manager.py", line 337, in traced_hookexec
    return outcome.get_result()
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\callers.py", line 52, in from_call
    result = func()
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\manager.py", line 335, in <lambda>
    outcome = _Result.from_call(lambda: oldcall(hook, hook_impls, kwargs))
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "c:\users\daniel\documents\spyder\spyder\external-deps\python-language-server\pyls\plugins\symbols.py", line 12, in pyls_document_symbols
    definitions = document.jedi_names(all_scopes=all_scopes)
  File "c:\users\daniel\documents\spyder\spyder\external-deps\python-language-server\pyls\workspace.py", line 221, in jedi_names
    script = self.jedi_script()
  File "c:\users\daniel\documents\spyder\spyder\external-deps\python-language-server\pyls\workspace.py", line 234, in jedi_script
    environment = self.get_enviroment(environment_path) if environment_path else None
  File "c:\users\daniel\documents\spyder\spyder\external-deps\python-language-server\pyls\workspace.py", line 259, in get_enviroment
    environment = jedi.api.environment.create_environment(path=environment_path, safe=False)
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\jedi\api\environment.py", line 376, in create_environment
    return Environment(path)
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\jedi\api\environment.py", line 67, in __init__
    self._get_subprocess()
  File "C:\Users\Daniel\Anaconda3\envs\develop\lib\site-packages\jedi\api\environment.py", line 80, in _get_subprocess
    exc))

jedi.api.environment.InvalidPythonEnvironment: Could not get version information for 'C:/Users/Daniel/Documents/Quansight/gimp-2.10.14-setup-3.exe': InternalError("The subprocess C:/Users/Daniel/Documents/Quansight/gimp-2.10.14-setup-3.exe has crashed (EOFError('Ran out of input'), stderr=).")
```

Maybe this was already an error with an issue but just in case...

**Comment by goanpeca** [comment]
>This happens not only in this PR but also in master and 4.x, although in 4.x no error report raises (but you can see that under the hood something is trying to execute the path given). I tested passing as interpreter a path to an installer (GIMP installer in this case): ?

@dalthviz could you open a separate issue

**Comment by goanpeca** [comment]
This is ready for review @ccordoba12

File: spyder/plugins/maininterpreter/confpage.py [hunk 1]
```diff
@@ -191,26 +213,39 @@ def set_umr_namelist(self):
...
+                        _('UMR'),
+                        _("The following modules are not "
+                          "installed on your machine:\n%s") % invalid,
+                        QMessageBox.Ok,
+                    )
+                QMessageBox.information(
+                    self,
+                    _('UMR'),
+                    _("Please note that these changes will "
+                      "be applied only to new Python/IPython "
```
Comment (ccordoba12) [inline spyder/plugins/maininterpreter/confpage.py:242]: Let's remove `Python` from here. It's not necessary anymore.

File: spyder/plugins/maininterpreter/plugin.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            self.set_conf_option(
+                'executable',
+                executable.replace("pythonw.exe", "python.exe"))
+
+        if not self.get_conf_option('default'):
+            if not self.get_conf_option('custom_interpreter'):
+                self.set_conf_option('custom_interpreter', ' ')
+
+    # --- Public API
+    # ------------------------------------------------------------------------
```
Comment (ccordoba12) [inline spyder/plugins/maininterpreter/plugin.py:78]: If this block comment is not necessary, please remove it.

File: spyder/utils/tests/test_programs.py [hunk 3]
```diff
@@ -98,8 +98,9 @@ def test_run_python_script_in_terminal(scriptpath, qtbot):
 @flaky(max_runs=3)
 @pytest.mark.first
 @pytest.mark.skipif(
-    os.environ.get('CI', None) is None or os.name == 'nt',
-    reason='Only on CI and not on windows!',
+    (os.environ.get('CI', None) is None
+     or (sys.platform.startswith("linux") or os.name == "nt")),
+    reason='Only on CI and only mac!',
```
Comment (ccordoba12) [inline spyder/utils/tests/test_programs.py:103]: Please revert this. It was caused by not installing a terminal on the Linux CIs. But I solved that several weeks ago.

File: spyder/utils/tests/test_programs.py [hunk 4]
```diff
@@ -122,8 +123,9 @@ def test_run_python_script_in_terminal_blank_wdir(scriptpath_with_blanks,
...
```
Comment (ccordoba12) [inline spyder/utils/tests/test_programs.py:128]: Please also revert this.

**Comment by ccordoba12** [comment]
I tested this manually and it's working as expected. So other than the small comments I left above, this one is ready.

**Review by ccordoba12 — APPROVED** [review]
This is ready. Thanks @goanpeca!

**Pull request merged**